modin/core/dataframe/pandas/partitioning/axis_partition.py

```diff
@@ -155,12 +155,14 @@ def deploy_axis_func(
         # Pop these off first because they aren't expected by the function.
         manual_partition = kwargs.pop("manual_partition", False)
         lengths = kwargs.pop("_lengths", None)
+        impure = kwargs.pop("_impure", False)
```

Comment:
what does `impure` mean?

Reply:
Whenever we run a function on a partition, we wrap the results in a partition object. This is fine for block partitions, but for axis partitions, that results in us trying to split back up the result, which errors out since it tries to get lengths along different axes, which is unsupported for some of the objects returned by the functions we're using here, so `impure` is basically my hacky way of telling the axis partition that it should just return the object directly without trying to split results.

Reply:
In that case wouldn't it be more `pure` than `impure`? I guess I'm just arguing semantic differences :D

Reply:
good point - I'm not entirely sure myself. To begin with, this is kind of a hacky workaround, so I'm not sure we want to add it to the codebase (thoughts @devin-petersohn?), but my reasoning was that if it were "pure" it would return an output of the same type as its input (since that's what apply expects) but since its not doing that, it's "impure"?

Reply:
@RehanSD  it looks like you've introduced `_impure` to accommodate applying axis-wide `sample_func`, which returns both a full-1-axis pandas dataframe and a list of quantiles for the first column of that dataframe. `_impure` creates a very different alternate API for all the axis-wide functions in the partition manager layer. With `_impure=True`, the per-partition function can map from a pandas dataframe to any iterable object, and the entire function maps from lists of valid pandas dataframe partitions (i.e., each partition's `_data_` is a pandas dataframe) to lists of (possibly invalid) partitions (e.g. `PandasOnRayDataFramePartition`) of anything (you are creating a row of partitions each wrapping `np.array` of quantiles instead of pandas dataframes). I think this behavior is likely to confuse users of `map_axis_partitions`, and it's tricky to use even if you want it.

Would it work to instead crate row partitions with a single element by deploying an axis-wide identity function with `num_splits=1`? We could also create a pandas dataframe with a single row to represent the quantiles, so we always have valid partitions in the end.

Reply:
I need the first step to accomplish two things - to materialize the entire full axis partition with one physical block partition, and to return a sample of that block partition. There are four ways I believe I can accomplish this:
1. Use the existing map axis mechanism, specify lengths so that one block partition is materialized, and attach the samples as an additional column to the partition. This is not ideal, since the next step requires materializing the samples on the head node, which will require us to materialize each partition and then remove the column. 
2. Use the existing map axis partitions to generate the sample, and keep the axis partitioned block partitioned at the physical layer - causing us to need to do a full axis join a second time in the final, splitting, stage (preventing us from amortizing the join cost across the sample and split steps). 
3. Use the existing map axis partitions twice - once to generate samples, and once to materialize the axis partition as one block partition - this will add overhead since we're literally adding a dummy call just to materialize the axis partition as one block partition. 
4. Create a new mechanism that allows users to manually partition the axis partition from the remote function, and to respect that partitioning when wrapping results at the end - allowing me to transform the axis partition into two block partitions - one containing the entire partition materialized, and the other containing the sample. 

I'm currently leaning towards Option 4 - it's what I tried to accomplish (in a hacky way) with the `_impure` argument, but I plan to create a formal mechanism to do so. Thoughts?

Reply:
@mvashishtha is right that this breaks the implicit contract that the `_data` member of the block partition be a pandas dataframe but there is precedence - this is what we do when materializing block dimensions. Additionally, while it is a standard assumption that the data in a block partition is a pandas dataframe, there isn't any code that should break as long as this partition (the one backed by a sample) isn't passed around outside this code, since there's no automatic mechanisms in the partition code that rely on the data being a pandas dataframe (like there are for the axis partition's code)

Reply:
@RehanSD I agree that if we want to get the samples and the axis=1 full-axis partitions with a single remote call, (4) is the best solution. However, I'm still uncomfortable with (4) because getting two results out of a single call seems like a very particular use case of `apply_full_axis` that no one else will want. Really what you are doing here is pipelining two remote steps and also getting the result of each. The first step is effectively making full-axis-1 virtual partitions and `force_materialization()`ing them, and the other is applying a sampling function to each axis partition.

Is it possible to use the batch API to pipeline the two functions and only materialize the results of the sampling, then use the axis partition results from the first step? Otherwise, can you think of some other clean way to combine the functions and get the sample out with a single remote get?

In the absence of a cleaner solution, I can accept solution (4) for its performance benefit, but you should add a note explaining why we are doing this.

Reply:
@RehanSD I happened to be thinking about this some more, and I think we should go with your solution (2) for now. I have a few reasons.

a) Serialization/deserialization is really expensive, but the concat is not terribly so. This optimization would save on the concat, but would not save any serialization/deserialization. Even for this 2 GB pandas dataframe, pandas takes 2.4 sec to sort the entire dataframe by 1 column, but only .103 sec to concatenate one row of partitions:

```python
import pandas as pd
import numpy as np

num_axis_partitions = int(2**4)
partitions = [[pd.DataFrame(np.random.randint(0,100,(2**16,2**4))) for _ in range(num_axis_partitions)] for _ in range(num_axis_partitions)]
%time first_row = pd.concat(partitions[0], axis=1)
df = pd.concat([pd.concat([p for p in row], axis=1, ignore_index=True) for row in partitions], ignore_index=True)
%time sorted = df.sort_values(0)
```

b) If you want to cut out the time for the extra concat, I think a more natural solution is to have the virtual partition `apply` method update its list of blocks to only contain the concatenated dataframe. We should do that in a separate PR. But this is something we should think about after we've profiled performance and observed that the `concat` is taking a long time.
c) in the very long term (maybe 6 months from now), we should be pipelining these steps anyway, so 
d) I am still uncomfortable with solution (4) because I think it's kind of awkward and it would only be useful for this use case.

Reply:
This makes a lot of sense to me - I've rewritten sort to do this now!